File: Books/JavaRLafore/Stack/Java/stackList.java
package Books.Stack.Java;

class StackListApp {
    public static void main(String[] args) {

        StackList theStack = new StackList();

        theStack.push(10);
        theStack.push(30);
        theStack.push(50);

        theStack.displayStack();

        theStack.push(70);
        theStack.push(90);
        theStack.push(110);

        theStack.displayStack();

        theStack.pop();
        theStack.pop();
        theStack.pop();

        theStack.displayStack();
    }
}

class Link {
    public long dData;
    public Link next;

    public Link(long data) {
        dData = data;
    }

    public void displayLink() {
        System.out.print("{data: " + dData + "} ");
    }

}

class LinkedList {
    private Link first;

    public LinkedList() {
        first = null;
    }

    public boolean isEmpty() {
        return first == null;
    }

    public void insertFirst(long data) {
        Link newLink = new Link(data);
        newLink.next = first;
        first = newLink;
    }

    public long deleteFirst() {
        if (isEmpty()) {
            return -1;
        }

        Link temp = first;
        first = first.next;
        return temp.dData;
    }

    public void displayList(){
        Link current = first;

        while(current != null) {
            current.displayLink();
            current = current.next;
        }
        System.out.println("");
    }
}

class StackList {
    private LinkedList theList;

    public StackList() {
        theList = new LinkedList();
    }

    public void push(long data) {
        theList.insertFirst(data);
    }
    
    public long pop() {
        return theList.deleteFirst();
    }

    public boolean isEmpty() {
        return theList.isEmpty();
    }

    public void displayStack() {
        System.out.print("Stack (top -> bottom): ");
        theList.displayList();
    }
}

File: Books/JavaRLafore/Stack/JS/stackList.js
class Node {
    dData;
    next;

    constructor(data) {
        this.dData = data;
        this.next = null;
    }

    displayNode() {
        console.log(`{data: ${this.dData}} `);
    }
}

class LinkedList {
    #first;

    constructor() {
        this.#first = null;
    }

    isEmpty() {
        return this.#first === null;
    }

    insertFirst(data) {
        const newNode = new Node(data);
        newNode.next = this.#first;
        this.#first = newNode
    }

    deleteFirst() {
        if (this.isEmpty()) {
            return -1;
        }

        const temp = this.#first;
        this.#first = this.#first.next;

        return temp;
    }

    displayList() {
        let current = this.#first;

        while(current !== null) {
            current.displayNode();
            current = current.next;
        }
    }
}

class StackList {
    #theStack;

    constructor() {
        this.#theStack = new LinkedList();
    }

    push(data) {
        this.#theStack.insertFirst(data);
    }
    
    pop() {
        return this.#theStack.deleteFirst();
    }

    isEmpty() {
        return this.#theStack.isEmpty();
    }

    displayStack() {
        console.log("Stack (top -> bottom): ")
        this.#theStack.displayList();
    }
}

(function StackListApp(){
    const theStack = new StackList();

    theStack.push(10);
    theStack.push(30);
    theStack.push(50);

    theStack.displayStack();

    theStack.push(70);
    theStack.push(90);
    theStack.push(110);

    theStack.displayStack();

    theStack.pop();
    theStack.pop();
    theStack.pop();

    theStack.displayStack();
})()